source/MaterialXCore/Material.h

```diff
@@ -19,7 +19,7 @@ MATERIALX_NAMESPACE_BEGIN
 /// filtered by the given shader type and target.  By default, all surface shader nodes
 /// are returned.
 /// @param materialNode The node to examine.
-/// @param nodeType THe shader node type to return.  Defaults to the surface shader type.
```

Comment:
@StefanHabel Starting with this edit, I'm seeing a great set of fixes to spelling, letter case, and other details in the C++ libraries for MaterialX.  Would it make sense to separate these out into their own pull request, so that they can be reviewed independently of the new Python Documentation functionality in this proposal?

Reply:
yes, thanks @jstone-lucasfilm !

This PR is getting a little unwieldy.

We could generally split this into separate PRs:

1. Base functionality of generating the Python API Documentation via Sphinx.
2. Fixes in header files for typos encountered while copying docs in headers to docstrings in the Python bindings.
3. Added argument names and docstrings in the Python bindings.

We could treat this PR here (1567) as the one for the Sphinx base functionality, splitting off parts 2 and 3 into separate PRs.

The third part (adding the actual docstrings) could be broken apart further into four separate PRs for the main (groups of) MaterialX Python modules:
- PyMaterialXCore
- PyMaterialXFormat
- PyMaterialXGenShader
    - PyMaterialXGenGlsl
    - PyMaterialXGenOsl
    - PyMaterialXGenMdl
    - PyMaterialXGenMsl
- PyMaterialXRender
    - PyMaterialXRenderGlsl
    - PyMaterialXRenderOsl
    - PyMaterialXRenderMsl

While working on this, I also encountered a couple of issues in the bindings themselves:

- `PyMaterialXCore.UnitDef.setUnitType()` is wrongly bound to `hasUnitType()`: https://github.com/AcademySoftwareFoundation/MaterialX/blob/main/source/PyMaterialX/PyMaterialXCore/PyDefinition.cpp#L76-L77
- `PyMaterialXCore.GeomPropDef`: duplicate definitions can be removed: https://github.com/AcademySoftwareFoundation/MaterialX/blob/main/source/PyMaterialX/PyMaterialXCore/PyGeom.cpp#L60-L71
    - `setGeomProp()`
    - `hasGeomProp()`
    - `getGeomProp()`
- `PyMaterialXGenShader.HwShaderGenerator` static functions are wrongly bound as instance methods: https://github.com/AcademySoftwareFoundation/MaterialX/blob/main/source/PyMaterialX/PyMaterialXGenShader/PyHwShaderGenerator.cpp#L31-L33
    - `bindLightShader()`
    - `unbindLightShader()`
    - `unbindLightShaders()`
- `PyMaterialXRenderMsl.MetalTextureHandler.bindImage()` is wrongly bound to `unbindImage()`: https://github.com/AcademySoftwareFoundation/MaterialX/blob/main/source/PyMaterialX/PyMaterialXRenderMsl/PyMetalTextureHandler.mm#L17-L18

We could treat these issues in the Python bindings as a separate ticket and PR.

Here's a list of tickets and corresponding PRs in which we could organize the overall work here:
- New issue ticket for existing issues in Python bindings
    - New PR to address the issues
- #342
    - #1567 for the Sphinx part
    - Separate new PR for docstrings for PyMaterialXCore
    - Separate new PR for docstrings for PyMaterialXFormat
    - Separate new PR for docstrings for PyMaterialXGen{Shader,Glsl,Osl,Mdl,Msl}
    - Separate new PR for docstrings for PyMaterialXRender{,Glsl,Osl,Msl}
    - Separate new PR for fixes in header files

Alternatively:
- New issue ticket for existing issues in Python bindings
    - New PR to address the issues
- #342
    - #1567 for the Sphinx part and docstrings in all Python modules
    - Separate new PR for fixes in header files

I'd be happy with either approach.

I have more commits to add to this PR here, adding more argument names and docstrings. I'll keeping using this PR here for this for now.

Reply:
@StefanHabel One approach that could be useful here is to break out one separable change at a time from this larger pull request (e.g. fixes to C++ libraries), so that it can be reviewed and merged to MaterialX main on its own.

As each of these subsets is reviewed and merged to main, we can then click "Update Branch" on this larger pull request, effectively removing that subset from the main proposal.

This should help to focus reviewer attention on a single, independent change at a time, rather than considering the entire proposal at once.

Reply:
Understood! Sounds good!

I'll upload a few more commits to this PR here, to have everything in one place.

I'll then create separate PRs for subsets of this work, which can be reviewed and merged independently.